packages/builtin/string.pony

```diff
@@ -93,56 +93,39 @@ class val String is (Seq[U8] & Comparable[String box] & Stringable)
     """
     Create a UTF-8 string from a single UTF-32 code point.
     """
-    if value < 0x80 then
-      _size = 1
-      _alloc = _size + 1
-      _ptr = Pointer[U8]._alloc(_alloc)
-      _set(0, value.u8())
-    elseif value < 0x800 then
-      _size = 2
-      _alloc = _size + 1
-      _ptr = Pointer[U8]._alloc(_alloc)
-      _set(0, ((value >> 6) or 0xC0).u8())
-      _set(1, ((value and 0x3F) or 0x80).u8())
-    elseif value < 0xD800 then
-      _size = 3
-      _alloc = _size + 1
-      _ptr = Pointer[U8]._alloc(_alloc)
-      _set(0, ((value >> 12) or 0xE0).u8())
-      _set(1, (((value >> 6) and 0x3F) or 0x80).u8())
-      _set(2, ((value and 0x3F) or 0x80).u8())
-    elseif value < 0xE000 then
-      // UTF-16 surrogate pairs are not allowed.
-      _size = 3
-      _alloc = _size + 1
-      _ptr = Pointer[U8]._alloc(_alloc)
-      _set(0, 0xEF)
-      _set(1, 0xBF)
-      _set(2, 0xBD)
-      _size = _size + 3
-    elseif value < 0x10000 then
-      _size = 3
-      _alloc = _size + 1
-      _ptr = Pointer[U8]._alloc(_alloc)
-      _set(0, ((value >> 12) or 0xE0).u8())
-      _set(1, (((value >> 6) and 0x3F) or 0x80).u8())
-      _set(2, ((value and 0x3F) or 0x80).u8())
-    elseif value < 0x110000 then
-      _size = 4
-      _alloc = _size + 1
-      _ptr = Pointer[U8]._alloc(_alloc)
-      _set(0, ((value >> 18) or 0xF0).u8())
-      _set(1, (((value >> 12) and 0x3F) or 0x80).u8())
-      _set(2, (((value >> 6) and 0x3F) or 0x80).u8())
-      _set(3, ((value and 0x3F) or 0x80).u8())
-    else
-      // Code points beyond 0x10FFFF are not allowed.
-      _size = 3
-      _alloc = _size + 1
-      _ptr = Pointer[U8]._alloc(_alloc)
-      _set(0, 0xEF)
-      _set(1, 0xBF)
-      _set(2, 0xBD)
+    let encoded = _UTF32Encoder.encode(value)
+    _size = encoded._1
+    _alloc = _size + 1
+    _ptr = Pointer[U8]._alloc(_alloc)
+    _set(0, encoded._2)
+    if encoded._1 > 1 then
+      _set(1, encoded._3)
+      if encoded._1 > 2 then
+        _set(2, encoded._4)
+        if encoded._1 > 3 then
+          _set(3, encoded._5)
+        end
+      end
+    end
+    _set(_size, 0)
+
+  fun ref push_utf32(value: U32) =>
+    """
+    Push a UTF-32 code point.
+    """
+    let encoded = _UTF32Encoder.encode(value)
+    let i = _size
+    _size = _size + encoded._1
+    reserve(_size)
+    _set(i, encoded._2)
+    if encoded._1 > 1 then
+      _set(i + 1, encoded._3)
+      if encoded._1 > 2 then
+        _set(i + 2, encoded._4)
+        if encoded._1 > 3 then
+          _set(i + 3, encoded._5)
+        end
+      end
     end
```

Comment:
It seems like there's still a lot of duplicate code here.  Would it be good to encapsulate these four `_set` calls behind a shared definition?  Seems like you might not even need the tuple return from `_UTF32Encoder` in that case as well (or maybe just a smaller-sized tuple).

I'm not 100% sure that what I suggested above is a good idea, but I wanted to bring up the fact that you refactored these two methods to remove a lot of otherwise-duplicate code, but there is still a significant chunk of duplicate code that *could* possibly be removed in some other refactor.  Just food for thought.